### PR TITLE
Fix deadlocks [Rebased, Squashed]

### PR DESCRIPTION
**Pull request by fabianhjr**
This is a rebase + squashes of @Russoul and @abailly work to fix deadlocks.

Supersedes #1298

### CODE REVIEW

File: src/Compiler/Scheme/Chez.idr [hunk 1]
```diff
@@ -282,19 +312,25 @@ schemeCall fc sfn argns ret
...
            Just ("scheme", [sfn]) =>
                do body <- schemeCall fc sfn (map fst args) ret
                   pure ("", body)
-           Just ("C", [cfn, clib]) => cCall appdir fc cfn clib args ret
-           Just ("C", [cfn, clib, chdr]) => cCall appdir fc cfn clib args ret
-           _ => throw (NoForeignCC fc)
+           Just ("C__collect_safe", (cfn :: clib :: _)) => do
+             when (unsupportedCallingConvention version) $
+               throw (GenericMsg fc $ "Chez version " ++ show version ++  " is undefined or does not support '__collect_safe' calling convention")
+             cCall appdir fc cfn clib args ret True
```
Comment (fabianhjr) [inline src/Compiler/Scheme/Chez.idr:328]: Here is the a check of support by @andorp

Comment (gallais) [inline src/Compiler/Scheme/Chez.idr:328]: Wasn't the plan to keep generating code without the collect safe calling convention if
the available scheme wasn't recent enough?

Comment (fabianhjr) [reply src/Compiler/Scheme/Chez.idr:328]: I think I could create an extra commit on top of Russoul's and Abailly's work to change the detection from an error if unsupported to switch to the safe on collect if supported.

Comment (fabianhjr) [reply src/Compiler/Scheme/Chez.idr]: Changed to:

```
             if unsupportedCallingConvention version
               then cCall appdir fc cfn clib args ret False
               else cCall appdir fc cfn clib args ret True
```

Comment (fabianhjr) [reply src/Compiler/Scheme/Chez.idr]: Ended up amending rather than creating an extra commit.

**Comment by abailly-iohk** [comment]
Thanks @fabianhjr for following up on this!

**Comment by fabianhjr** [comment]
You are welcome, thanks a lot for your work! UwU

**Pull request merged**